modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/tests/src/Functional/SectionRulesTest.php

```diff
@@ -0,0 +1,60 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Drupal\Tests\oe_webtools_analytics_rules\Functional;
+
+use Drupal\Core\Cache\Cache;
+use Drupal\oe_webtools_analytics\AnalyticsEventInterface;
+use Drupal\Tests\BrowserTestBase;
+
+/**
+ * Class SectionRulesTest.
+ *
+ * @group oe_webtools_analytics
+ */
+class SectionRulesTest extends BrowserTestBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  public static $modules = [
+    'config',
+    'system',
+    'oe_webtools',
+    'oe_webtools_analytics',
+    'oe_webtools_analytics_rules',
+  ];
+
+  /**
+   * Test that the Webtools JavaScript library is correctly loaded on a page.
+   */
+  public function testLibraryLoading(): void {
+    $config = \Drupal::configFactory()
+      ->getEditable(AnalyticsEventInterface::CONFIG_NAME)
+      ->set("siteID", "123")
+      ->set("sitePath", "ec.europa.eu");
+    $config->save();
+
+    $this->container->get('entity_type.manager')
+      ->getStorage('webtools_analytics_rule')
+      ->create(['id' => 'id1', 'section' => 'section1', 'regex' => '/admin/'])
+      ->save();
+
+    foreach (Cache::getBins() as $service_id => $cache_backend) {
+      if ('dynamic_page_cache' === $service_id || 'page' === $service_id) {
+        $cache_backend->deleteAll();
```

Comment:
Why is this being done here? Does this perhaps mean the functionality only works if the user manually clears the caches after creating a new rule? That seems like a bug to me that is probably worth fixing.

I'm not sure that hardcoding the two page caching backends that are included in core is the recommended way. There might be more caching mechanisms. The right way to handle this is by creating a cache context.

Reply:
To be fair I based the test on Drupal\Tests\oe_webtools_analytics\Kernel\ConfigurationTest and I didn't really think about it. Do you have any documentation on how to manage cache contexts?

Reply:
OK so what we need is a way to tell Drupal: "Some of our pages will have webtools JS inserted into them, and whether or not this is the case is depending on this dynamic rule engine in my module. The rules may change at any time, depending on certain factors. In fact we can have two versions of any page: one with the JS inserted, and one without".

In Drupal 7 we didn't have a way to declare whether or not some page cache should be invalidated depending on certain conditions. The only way to make it work was to clear the entire page cache when the user updates the rules.

But now in Drupal 8 we have cache contexts! We can say things like "This page will be rendered differently if the root user is logged in" -> `IsSuperUserCacheContext`. Or "I am showing dates and times on the page, and this is rendered differently for different time zones" -> `TimeZoneCacheContext`.

In our case it is "Depending on whether or not our dynamic rule engine determines that the current page belongs to a certain section we will render the page differently (the JS will have a different value for the `siteSection`)" -> `OpenEuropaWebtoolsAnalyticsSiteSectionCacheContext`!

You just have to write a simple class that checks to which site section the current page belongs and return it. I would suggest to use `WorkspaceCacheContext` as an example since this is a very similar use case. You'll also need to declare the service (check `workspace.services.yml`).

The official documentation is here: https://www.drupal.org/docs/8/api/cache-api/cache-contexts